base/staticint.jl

```diff
@@ -0,0 +1,133 @@
+Base.promote_rule(::Type{T1}, @nospecialize(T2::Type{<:StaticInt})) where {T1>:Union{Missing,Nothing}} = promote_type(T1, Int)
+Base.promote_rule(::Type{T1}, @nospecialize(T2::Type{<:StaticInt})) where {T1>:Nothing} = promote_type(T1, Int)
+Base.promote_rule(::Type{T1}, @nospecialize(T2::Type{<:StaticInt})) where {T1>:Missing} = promote_type(T1, Int)
+for T in [:Bool, :Missing, :BigFloat, :BigInt, :Nothing, :Any]
+    # let S = :Any
+    @eval begin
+        Base.promote_rule(@nospecialize(S::Type{<:StaticInt}), ::Type{$T}) = promote_type(Int, $T)
+        Base.promote_rule(::Type{$T}, @nospecialize(S::Type{<:StaticInt})) = promote_type($T, Int)
+    end
+end
+Base.promote_rule(@nospecialize(T1::Type{<:StaticInt}), @nospecialize(T2::Type{<:StaticInt})) = Int
+
+Base.:(%)(@nospecialize(n::StaticInt), ::Type{Integer}) = Int(n)
+
+Base.eltype(@nospecialize(T::Type{<:StaticInt})) = Int
+Base.iszero(::Zero) = true
+Base.iszero(@nospecialize(x::StaticInt)) = false
+Base.isone(::One) = true
+Base.isone(@nospecialize(x::StaticInt)) = false
+Base.zero(@nospecialize(x::Type{<:StaticInt})) = Zero()
+Base.one(@nospecialize(x::Type{<:StaticInt})) = One()
+
+for T in [:Real, :Rational, :Integer]
+    @eval begin
+        @inline Base.:(+)(x::$T, y::Zero) = x
+        @inline Base.:(+)(x::$T, @nospecialize(y::StaticInt)) = x + Int(y)
+        @inline Base.:(+)(x::Zero, y::$T) = x
+        @inline Base.:(+)(@nospecialize(x::StaticInt), y::$T) = Int(x) + y
+        @inline Base.:(-)(x::$T, y::Zero) = x
+        @inline Base.:(-)(x::$T, @nospecialize(y::StaticInt)) = x - Int(y)
+        @inline Base.:(*)(x::$T, y::Zero) = Zero()
+        @inline Base.:(*)(x::$T, y::One) = x
+        @inline Base.:(*)(x::$T, @nospecialize(y::StaticInt)) = x * Int(y)
+        @inline Base.:(*)(x::Zero, y::$T) = Zero()
+        @inline Base.:(*)(x::One, y::$T) = y
+        @inline Base.:(*)(@nospecialize(x::StaticInt), y::$T) = Int(x) * y
+    end
+end
+@inline Base.:(+)(::Zero, ::Zero) = Zero()
+@inline Base.:(+)(::Zero, ::StaticInt{M}) where {M} = StaticInt{M}()
+@inline Base.:(+)(::StaticInt{M}, ::Zero) where {M} = StaticInt{M}()
+
+@inline Base.:(-)(::StaticInt{M}) where {M} = StaticInt{-M}()
+@inline Base.:(-)(::StaticInt{M}, ::Zero) where {M} = StaticInt{M}()
+
+@inline Base.:(*)(::Zero, ::Zero) = Zero()
+@inline Base.:(*)(::One, ::Zero) = Zero()
+@inline Base.:(*)(::Zero, ::One) = Zero()
+@inline Base.:(*)(::One, ::One) = One()
+@inline Base.:(*)(::StaticInt{M}, ::Zero) where {M} = Zero()
+@inline Base.:(*)(::Zero, ::StaticInt{M}) where {M} = Zero()
+@inline Base.:(*)(::StaticInt{M}, ::One) where {M} = StaticInt{M}()
+@inline Base.:(*)(::One, ::StaticInt{M}) where {M} = StaticInt{M}()
+for f in [:(+), :(-), :(*), :(/), :(÷), :(%), :(<<), :(>>), :(>>>), :(&), :(|), :(⊻)]
+    @eval @generated function Base.$f(::StaticInt{M}, ::StaticInt{N}) where {M,N}
+        return Expr(:call, Expr(:curly, :StaticInt, $f(M, N)))
+    end
+end
```

Comment:
Couldn't all these definitions just be written as:

```jl
for f in [:(+), :(-), :(*), :(/), :(÷), :(%), :(<<), :(>>), :(>>>), :(&), :(|), :(⊻)]
   @eval begin
      @inline Base.$f(::StaticInt{M}, ::StaticInt{N}) where {M,N} = StaticInt{$f(M,N)}()
   end
end
```

I.e., what is the advantage of spelling out the `Zero` and `One` cases and why are generated functions needed?

Coincidentally, should `/` be defined for `StaticInt`s in this way (integers are not closed under division)?

Reply:
I'm actually not sure why we had those explicitly defined.
The generated function is left over from work trying to support LoopVectorization but should be removed and I agree that `/` should probably not be there.